ksmt-core/src/main/kotlin/org/ksmt/cache/weak/WeakHashMapCache.kt

```diff
@@ -0,0 +1,98 @@
+package org.ksmt.cache.weak
+
+import org.ksmt.cache.CacheRemoveHandler
+import java.lang.ref.Reference
+import java.lang.ref.ReferenceQueue
+
+abstract class WeakHashMapCache<K : Any, V : Any> {
+    private val data = HashMap<Any, KeyRefNode<K, V>>()
+    private val keyReferenceQueue = ReferenceQueue<K>()
+    private var modificationsSinceLastCleanup = 0
+    private var removeHandler: CacheRemoveHandler<K, V>? = null
+
+    abstract fun lookupKey(key: K): Any
+
+    abstract fun newNode(key: K, referenceQueue: ReferenceQueue<K>, value: V, hash: Int): KeyRefNode<K, V>
```

Comment:
Looks the same as for `ConcurrentWeakHashMap`, why they don't have a common interface?

Reply:
They are not same, because `newNode` returns different types for `WeakHashMapCache.KeyRefNode` and `ConcurrentWeakHashMapCache.KeyRefNode`.